src/software/jetson_nano/services/motor.cpp

```diff
@@ -93,9 +125,137 @@ std::unique_ptr<TbotsProto::DriveUnitStatus> MotorService::poll(
+// the same SPI interface. So when we bind the API, we can use the same
+// readWriteByte function, provided that the chip select pin is turning on
+// the right chip.
+//
+// Each TMC4671 and TMC6100 pair have their chip selects coming in from a
+// demux (see diagram below). The demux is controlled by the
+// spi_cs_driver_to_controller_demux_gpio. When low, the chip select is passed
+// to the TMC4671, and to the TMC6100 when high.
+//
+//
+//                                               FRONT LEFT MOTOR
+//                                              CONTROLLER + DRIVER
+//                                             ┌─────────────────┐
+//                                             │                 │
+//                            ┌───────┐        │   ┌─────────┐   │
+//                            │       │ SEL(LOW)   │         │   │
+//                            │  1:2  ├────────┬───►TMC4671  │   │
+//                            │       │        │   └─────────┘   │
+//             FRONT_LEFT_CS  │ DEMUX │        │                 │
+//             ───────────────►       │        │   ┌─────────┐   │
+//                            │       │SEL(HIGH)   │         │   │
+//                            │       ├────────┬───►TMC6100  │   │
+//                            │       │        │   └─────────┘   │
+//                            │       │        │                 │
+//                            └───▲───┘        └─────────────────┘
+//                                │
+//                                │ SEL
+//                                │
+//                  spi_cs_driver_to_controller_demux
+//
+uint8_t MotorService::tmc4671ReadWriteByte(uint8_t motor, uint8_t data,
+                                           uint8_t last_transfer)
+{
+    spi_cs_driver_to_controller_demux_gpio.setValue(GpioState::LOW);
+    return readWriteByte(motor, data, last_transfer);
+}
+
+uint8_t MotorService::tmc6100ReadWriteByte(uint8_t motor, uint8_t data,
+                                           uint8_t last_transfer)
+{
+    spi_cs_driver_to_controller_demux_gpio.setValue(GpioState::HIGH);
+    return readWriteByte(motor, data, last_transfer);
+}
+
+uint8_t MotorService::readWriteByte(uint8_t motor, uint8_t data, uint8_t last_transfer)
+{
+    uint8_t ret_byte = 0;
+
+    if (!transfer_started)
+    {
+        memset(tx, 0, sizeof(tx));
+        memset(rx, 0, sizeof(rx));
+        position = 0;
+
+        if (data & TMC_WRITE_BIT)
+        {
+            // If the transfer started and its a write operation,
+            // set the appropriate flags.
+            currently_reading = false;
+            currently_writing = true;
+        }
+        else
+        {
+            // The first byte should contain the address on a read operation.
+            // Trigger a transfer (1 byte) and buffer the response (4 bytes)
+            tx[position] = data;
+            spiTransfer(file_descriptors[motor], tx, rx, 5);
```

Comment:
why read 4 bytes when the method returns one at a time? is this what API requires?

Reply:
yeah this is my gripe with it too, but i think its because its designed for microcontrollers

```
This is the APi function that calls this one:

int32_t tmc4671_readInt(uint8_t motor, uint8_t address)
{
	// clear write bit
	address &= 0x7F;

	// write address
	tmc4671_readwriteByte(motor, address, false);

	// read data
	int32_t value = tmc4671_readwriteByte(motor, 0, false);
	value <<= 8;
	value |= tmc4671_readwriteByte(motor, 0, false);
	value <<= 8;
	value |= tmc4671_readwriteByte(motor, 0, false);
	value <<= 8;
	value |= tmc4671_readwriteByte(motor, 0, true);

	return value;
}
```

Reply:
interesting